core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -734,6 +734,10 @@ private[kafka] class Acceptor(val endPoint: EndPoint,
         val endThrottleTimeMs = e.startThrottleTimeMs + e.throttleTimeMs
         throttledSockets += DelayedCloseSocket(socketChannel, endThrottleTimeMs)
         None
+      case e: IOException =>
+        info(s"Encounter IOException", e)
+        closeSocket(socketChannel)
```

Comment:
We might have to use `close(endPoint.listenerName, socketChannel)` here. Otherwise, we would not decrement the number of connections in the `connectionQuotas`. Could you double check?

Reply:
@dajac You're right. I've updated the JIRA issue descriptions. I have a new commit using `close(endPoint.listenerName, socketChannel)`.
For the test, since it involves injecting an IOException, I'm going to use Byteman or AspectJ. Or do you have any suggestion?

Reply:
Hmm... Good question. I don't think that we already use those library in the code base so it would be great if we could do it with what we already use. I am not sure if it is possible.

Reply:
I think Byteman is widely used in many other communities, because it's a very elegant way to do fault injection testing. For example, ZooKeeper community has been discussing it in https://github.com/apache/zookeeper/pull/123. I think adding Byteman dependency won't take much time.

Reply:
Another way would be to extract `serverSocketChannel.accept()` into a method. Then we could override that method in a test and return a mocked socket which throws an exception.

Reply:
It seems doable. I'm looking for a unit test in `core/src/test/scala/unit/kafka/network/SocketServerTest.scala` based on which I can modify & write the test. Do you have any suggestion?

Reply:
`testMaxConnectionsPerIpOverrides` might be a good starting point as it test something in the same area. Why do you think that we need to modify a log of code in `SocketServer`?

If we extract `serverSocketChannel.accept()` into a method. Then, we can create a `SocketServer` and overwrite the method. In the method, we could return a mocked Socket which throws an exception when `configureBlocking` is called. Then, we could try to establish a connection. Finally, we can verify that the `close` methods were called.

```scala
val overrideServer = new SocketServer(...) {
  override def accept(): Socket = {

  }
}
```

Reply:
There are multiple classes defined in `core/src/main/scala/kafka/network/SocketServer.scala`. The `accept` method belongs to `Acceptor`, rather than `SocketServer`, so we can't override `accept` the way you show. We need to override `accept` in `Acceptor`, and then use this `Acceptor` when the `SocketServer` is initialized. It seems none of the tests in `core/src/test/scala/unit/kafka/network/SocketServerTest.scala` have this kind of behavior. `Acceptor` appears multiple times within `SocketServer`, in multiple forms. So it seems we have to modify some code in `SocketServer` to enable our new `Acceptor`.

Reply:
You're right. I overlooked this. In this case, I wonder if we could test the Acceptor only. Would it work? That would simplify all of this.

Reply:
Since the existing test cases don't have scaffold for testing `Acceptor`, we need to prepare lots of parameters for `Acceptor` constructor
https://github.com/apache/kafka/blob/074a03cca162f91ccdecc12eb84c6a45af75f6bf/core/src/main/scala/kafka/network/SocketServer.scala#L550-L558
We also need to prepare a `Processor` for the `Acceptor`
https://github.com/apache/kafka/blob/074a03cca162f91ccdecc12eb84c6a45af75f6bf/core/src/main/scala/kafka/network/SocketServer.scala#L687-L703
https://github.com/apache/kafka/blob/074a03cca162f91ccdecc12eb84c6a45af75f6bf/core/src/main/scala/kafka/network/SocketServer.scala#L574-L578
https://github.com/apache/kafka/blob/074a03cca162f91ccdecc12eb84c6a45af75f6bf/core/src/main/scala/kafka/network/SocketServer.scala#L790-L806
However, those logic are implemented in `SocketServer`
https://github.com/apache/kafka/blob/074a03cca162f91ccdecc12eb84c6a45af75f6bf/core/src/main/scala/kafka/network/SocketServer.scala#L268-L283
https://github.com/apache/kafka/blob/074a03cca162f91ccdecc12eb84c6a45af75f6bf/core/src/main/scala/kafka/network/SocketServer.scala#L424-L445
Therefore, I think testing Acceptor requires much more code. Instead, testing `SocketServer` is simpler.